### PR TITLE
Fix issue 19873 - function should be by default @system even with -preview=dip1000

### PR DESCRIPTION
**Pull request by dkorpel**
Walter said in the issue:

> As you know, I'd like to see `@safe` as the default. While this issue is a bug, it's in the direction we need to go anyway so it shouldn't be "fixed".

However, we can't defy the specification based on hypothetical future language changes.

https://dlang.org/spec/function.html#scope-parameters
> `scope` escape analysis is only done for `@safe` functions. For other functions scope semantics must be manually enforced.

Currently D users should be able to assume two things:
- Functions are `@system` by default
- `@trusted` functions are like `@system` functions callable from `@safe` functions

These assumptions are subtly broken with dip1000 and it's very confusing when e.g. using `@trusted` to circumvent limitations of dip1000 results in errors anyways.

Note that the original example in the issue doesn't error anymore, but the test cases in this PR, demonstrating scope inference on assignment, still do.

### CODE REVIEW

**Review by MoonlightSentinel — APPROVED** [review]

**Comment by dkorpel** [comment]
This currently breaks `@safe` inference it seems

File: src/dmd/escape.d
Comment (MoonlightSentinel) [inline src/dmd/escape.d:593]: The issue is about invalid error messages in normal functions, so it might be sensible to check whether the safety is inferred for the current function.

**Review by MoonlightSentinel — DISMISSED** [review]
Needs work w.r.t. to templates

File: src/dmd/escape.d
Comment (dkorpel) [reply src/dmd/escape.d]: That's what I'm trying now, but it means this doesn't compile:
```
auto ad(scope int* x)
{
    auto y = x;
    return y; // error: returning scope variable y
}
```
I'm thinking maybe scope inference should be done always, but then checkReturnEscape/checkAssignEscape should not error when returning an inferred scope variable unless it's a `@safe` function. Or maybe it shouldn't at all, should this compile?
```D
int* f(scope int* x) @system 
{
    return x;
}
```
It currently doesn't with -dip1000, which arguably should be fixed too, unless you don't count that as "scope escape analysis" but simply the trivial case of directly returning a scope variable. After all, this isn't allowed, dip1000 or not:
```D
int* f(int x) @system 
{
    return &x; // Error: returning `& x` escapes a reference to local variable `x`
}
```

Comment (MoonlightSentinel) [reply src/dmd/escape.d]: The first and second examples are fine because they might be valid depending on their usage ("scope semantics must be manually enforced"). But the third example could be rejected because it's known to always produce a stale pointer.

**Review by RazvanN7 — APPROVED** [review]
Is this ready to go?

**Comment by dkorpel** [comment]
> Is this ready to go?

Kind of. I think it's an improvement on its own, but it leaves the question what to do about inferred functions and `@system` functions directly returning `scope` variables.

**Comment by RazvanN7** [comment]
As far as I understand, scope analysis should not apply for `system` functions. As for inferred functions, I think that the way to go is: first try to infer scope, then try to infer @safe.

File: src/dmd/escape.d
Comment (dkorpel) [reply src/dmd/escape.d]: I now removed the `inferScope` variable so inference happens always whether it's `@system`, default, interred, `@trusted` or `@safe`. Instead I added a `sc.func.setUnsafe()` check so it doesn't complain about returning `scope` variables in non-`@safe` code anymore.

**Comment by dkorpel** [comment]
CircleCI failure "Failed to download script install.sh" looks unrelated, it should be good to go now.

**Comment by dkorpel** [comment]
Ping @RazvanN7 @MoonlightSentinel

**Comment by MoonlightSentinel** [comment]
IIUC the current patch always infers scope but supresses error messages by checking for `preview=dip1000` and  `setUnsafe`?
That seems problematic for covariance checks / conversions of function literals (see `Parameter.isCovariantScope` used by `Type.covariant`), i.e. when `scope` is inferred unexpectedly.

Related: https://issues.dlang.org/show_bug.cgi?id=21285#c7

**Comment by dkorpel** [comment]
> IIUC the current patch always infers scope but supresses error messages by checking for `preview=dip1000` and `setUnsafe`? 

Kind of. It always infers `scope` on a variable when a `scope` variable is assigned to it, regardless of explicit `@system` or `@safe` annotations. It doesn't mean parameters will be inferred `scope` always, `FUNCFLAG.inferScope` is only set when `canInferAttributes` is true (meaning template function, `auto` return, function literals, the usual stuff).

I wouldn't call it 'surpression' of errors, most checks on scope variables (e.g. taking the address, passing to a function, assigning to a variable) are already behind `preview=dip1000` and `setUnsafe`. Only returning a `scope` variable is an exception, which is currently only behind a dip1000 check.

> That seems problematic for covariance checks / conversions of function literals (see `Parameter.isCovariantScope` used by `Type.covariant`), i.e. when `scope` is inferred unexpectedly.

Why is it problematic? Are people currently explicitly marking their function literals `@system` to make `scope` inference on parameters fail more easily?

**Comment by MoonlightSentinel** [comment]
> I wouldn't call it 'supression' of errors, most checks on scope variables (e.g. taking the address, passing to a function, assigning to a variable) are already behind `preview=dip1000` and `setUnsafe`. Only returning a `scope` variable is an exception, which is currently only behind a dip1000 check.

"Suppression" is probably a bit negative. The general idea was that the changes are performed eagerly while error messages are only emitted when the aformentioned conditions are met.

> Why is it problematic? Are people currently explicitly marking their function literals @system to make scope inference on parameters fail more easily?

That's probably a rare occurence but even function templates can explicitly be `@system`.

I'm more concerned that the current patch causes *more* scope inference. The inferred `scope` is ignored w.r.t. safety errors but still affects the type of the function [literal]. The the linked bug report contains an example where missmatched `scope` inference breaks an associative array literal. Adding more inference for otherwise unused `scope` makes such problems more likely.

I am also not sure why more inference is necessary to prevent invalid error messages in `@system` functions.

**Comment by RazvanN7** [comment]
> It always infers scope on a variable when a scope variable is assigned to it, regardless of explicit @system or @safe annotations. 

Why? Shouldn't this inference be done only in the context of dip1000 and @safe functions? I understand that for template functions this is problematic because safety also needs to be inferred but this can be fixed by checking whether the function is inside a template instance.

I agree with @MoonlightSentinel  that scope inference should not be done carelessly.

**Comment by dkorpel** [comment]
> I agree with @MoonlightSentinel that scope inference should not be done carelessly.

There's two things we're talking about here:
- inferring a function parameter `scope` when it doesn't escape
- inferring the `scope` storage class for `x` when you assign `int* x = scopeVar`

This PR affects the second one, not the first one. I'm trying to see if this makes a difference here:

```D
auto ff0 = (int* x0)         {};
auto ff1 = (int* x1) @system {};
auto ff2 = (int* x2)         {scope int* y = x2;};
auto ff3 = (int* x3) @system {scope int* y = x3;};
auto ff4 = (int* x4)         {int* y = x4;};
auto ff5 = (int* x5) @system {int* y = x5;};
void nonScope(int* x) {}
auto ff6 = (int* x6)         {nonScope(x6);};
auto ff7 = (int* x7) @system {nonScope(x7);};
```

With -preview=dip1000: Inferring scope for x0, x1, x2, x3
Without -preview=dip1000: Inferring scope for x0, x1, x2, x3, x6, x7

- `@system` doesn't affect anything here
- Without dip1000, ff6 and ff7 incorrectly infer `scope` on the parameter
- It's the same *both before and after* this PR. 

So what does this PR affect?
```D
auto fun(scope int* x) // parameter *explicitly* marked scope
{
    auto y = x; // therefore y 'inferred' scope in checkAssignEscape
    return y; // error, despite not a `@safe` function
}
```
The point is that adding `@system` to a function that's already `@system` shouldn't affect anything, but currently it does.

Yes, when the function is not `@safe` and has no inferred attributes you could set `inferScope = false`, but why? The code is now simpler with 5 fewer lines and 6 fewer && expressions.

Issue 21285 should be fixed, but it's unrelated to this PR as far as I see it.

**Comment by MoonlightSentinel** [comment]
> There's two things we're talking about here:
>
> - inferring a function parameter scope when it doesn't escape
> - inferring the scope storage class for x when you assign int* x = scopeVar
>
> This PR affects the second one, not the first one.

Indeed, but the second one also applies to assignments to parameters. `p8` will be inferred as `scope` with the proposed changes:

```d
auto ff8 = (int* p8) @system
{
    scope int* l8;
    p8 = l8;
};
```

**Comment by dkorpel** [comment]
> Indeed, but the second one also applies to assignments to parameters. `p8` will be inferred as `scope` with the proposed changes:

`p8` also infers `scope` without the proposed changes, regardless of `-preview=dip1000`.

**Comment by MoonlightSentinel** [comment]
> p8 also infers scope without the proposed changes, regardless of -preview=dip1000.

It does not depend on `-preview=dip1000`, but `STC.scope` is set for `p8` [here](https://github.com/dlang/dmd/blob/d9f8fcef0c213c5dd0609799bdf45e0faa4b1cfb/src/dmd/escape.d#L716). That line is only hit with your proposed patch - or is it set in another place as well?

**Comment by dkorpel** [comment]
> That line is only hit with your proposed patch - or is it set in another place as well?

In semantic3:

https://github.com/dlang/dmd/blob/7646103bdef64e933d6c656a9b164aef9306dff8/src/dmd/semantic3.d#L463-L464

Then 'maybeScope' can be removed by assignment to something non-scope. If that doesn't happen:

https://github.com/dlang/dmd/blob/7646103bdef64e933d6c656a9b164aef9306dff8/src/dmd/semantic3.d#L1311-L1318

That's also why without -dip1000, `scope` is sometimes inferred when it shouldn't. However, this PR is a self-contained fix for issue 19873, I don't think it should attempt to fix 21285 as well.

**Comment by MoonlightSentinel** [comment]
> In semantic3:

Thanks. I guess the impact should be minimal or maybe even non-existant for generic code.

> That's also why without -dip1000, scope is sometimes inferred when it shouldn't. However, this PR is a self-contained fix for issue 19873, I don't think it should attempt to fix 21285 as well.

Agreed.

**Comment by dkorpel** [comment]
@RazvanN7 Are you fine with the current patch now?

**Review by RazvanN7 — APPROVED** [review]

**Comment by dkorpel** [comment]
Can this be merged?

**Comment by RazvanN7** [comment]
I see that @MoonlightSentinel still has not accepted this PR. Since he is more knowledgeable in this topic, I am waiting for his approval.

**Pull request merged**
